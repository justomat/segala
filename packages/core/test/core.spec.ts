import type { AnyAction } from 'redux';
import { call, put } from 'redux-saga/effects';
import { expectSaga } from 'redux-saga-test-plan';
import { linear, retry } from '../src';

jest.mock('../src/lib/backoff', () => ({
  exponential: () => () => 0,
  linear: (ms) => (n) => ms * 2 * (n + 1),
}));

const dummyClient = jest.fn().mockResolvedValue('Resolved');

function* dummySaga({ payload }: AnyAction) {
  try {
    const result = yield call(dummyClient, 'https://example.com', { payload });

    yield put({
      type: 'DUMMY_SUCCESS',
      payload: { result },
    });
  } catch ({ message, status }) {
    yield put({
      type: 'DUMMY_FAILURE',
      payload: { message, status },
    });
    yield put({
      type: 'DUMMY_SHOW_ALERT',
      payload: { message, status, type: 'error' },
    });
  }
}

function* brokenSaga({ payload }: AnyAction) {
  // @ts-ignore
  return yield call(dummyClient, payload.should.broke.here);
}

function stopConditionValidator(value: any) {
  if (value?.type !== 'PUT') {
    return false;
  }

  const { type, payload } = value.payload.action;

  return type.endsWith('_FAILURE') && ![401, 404, 500].includes(payload.status);
}

describe('retry', () => {
  beforeEach(() => {
    jest.spyOn(console, 'error');
    // @ts-ignore
    global.console.error.mockImplementation(() => {});
  });

  afterEach(() => {
    // @ts-ignore
    global.console.error.mockRestore();
    dummyClient.mockReset();
    dummyClient.mockResolvedValue('Resolved');
  });

  const action = { payload: { key: 'value' }, type: 'DUMMY_REQUEST' };
  const expectedArgs = [
    'https://example.com',
    expect.objectContaining({ payload: action.payload }),
  ];
  const originalGenerator = dummySaga;

  describe('on success cases', () => {
    it('should not change flow', async () => {
      const retryableGenerator = retry(originalGenerator);

      const originalResult = await expectSaga(originalGenerator, action).run();
      const retryableResult = await expectSaga(
        retryableGenerator,
        action
      ).run();

      expect(originalResult.toJSON()).toEqual(retryableResult.toJSON());
    });

    it('should execute once and dispatch the success action', async () => {
      const retryableGenerator = retry(originalGenerator);

      const result = await expectSaga(retryableGenerator, action).run();

      expect(result.toJSON()).toMatchSnapshot();
      expect(dummyClient).toHaveBeenCalledTimes(1);
      expect(dummyClient).toHaveBeenCalledWith(...expectedArgs);
      expect(result.effects.put).toHaveLength(1);
      expect(result.effects.put[0].payload.action.type).toMatch(/_SUCCESS$/g);
    });
  });

  describe('on error cases', () => {
    it('should retry 3 times (default) and dispatch the actions', async () => {
      dummyClient.mockRejectedValue('Rejected');
      const retryableGenerator = retry(originalGenerator);

      const result = await expectSaga(retryableGenerator, action).run();

      expect(result.effects.put).toHaveLength(2);
      expect(result.effects.put[0].payload.action.type).toMatch(/_FAILURE$/g);
      expect(result.effects.put[1].payload.action.type).toMatch(
        /^DUMMY_SHOW_ALERT$/g
      );
    });

    it('should retry 4 times', async () => {
      jest.setTimeout(10000);
      dummyClient.mockRejectedValue('Rejected');
      const retries = 4;
      const retryableGenerator = retry(originalGenerator, {
        backoff: linear(200),
        retries,
      });

      const result = await expectSaga(retryableGenerator, action).run();

      expect(result.toJSON()).toMatchSnapshot();
      expect(dummyClient).toHaveBeenCalledTimes(1 + retries); // original call + n retries
      expect(dummyClient.mock.calls).toEqual(
        Array(1 + retries).fill(expectedArgs)
      );
      expect(result.effects.call).toHaveLength(2 * retries + 1); // call(dummyClient) n+1 times + call(delay) n times
    });

    it('should retry 4 times using the meta from the action', async () => {
      dummyClient.mockRejectedValue('Rejected');
      const retryableGenerator = retry(originalGenerator);

      const result = await expectSaga(retryableGenerator, {
        ...action,
        meta: { retries: 4 },
      }).run();

      expect(result.toJSON()).toMatchSnapshot();
      expect(dummyClient).toHaveBeenCalledTimes(5);
    });

    it('should work with custom shouldAborts', async () => {
      dummyClient.mockRejectedValue({ message: 'Rejected', status: 401 });
      const retryableGenerator = retry(originalGenerator, {
        shouldAbort: stopConditionValidator,
      });
      const result = await expectSaga(retryableGenerator, action).run();

      expect(dummyClient).toHaveBeenCalledTimes(1);
      expect(result.toJSON()).toMatchSnapshot();
    });

    it('should not handle exceptions', async () => {
      const retryableGenerator = retry(brokenSaga);

      let result: any;
      let error: any;

      try {
        result = await expectSaga(retryableGenerator, action).run();
      } catch (e) {
        error = e;
      }

      expect(result).toBeUndefined();
      expect(error).toBeInstanceOf(TypeError);
    });
  });
});
