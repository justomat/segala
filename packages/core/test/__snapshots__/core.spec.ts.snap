// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`retry on error cases should retry 4 times 1`] = `
Object {
  "call": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          "https://example.com",
          Object {
            "payload": Object {
              "key": "value",
            },
          },
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/mockConstructor",
      },
      "type": "CALL",
    },
  ],
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "action": Object {
          "payload": Object {
            "message": undefined,
            "status": undefined,
          },
          "type": "DUMMY_FAILURE",
        },
        "channel": undefined,
      },
      "type": "PUT",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "action": Object {
          "payload": Object {
            "message": undefined,
            "status": undefined,
            "type": "error",
          },
          "type": "DUMMY_SHOW_ALERT",
        },
        "channel": undefined,
      },
      "type": "PUT",
    },
  ],
}
`;

exports[`retry on error cases should retry 4 times using the meta from the action 1`] = `
Object {
  "call": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          "https://example.com",
          Object {
            "payload": Object {
              "key": "value",
            },
          },
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/mockConstructor",
      },
      "type": "CALL",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          0,
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/delayP",
      },
      "type": "CALL",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          "https://example.com",
          Object {
            "payload": Object {
              "key": "value",
            },
          },
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/mockConstructor",
      },
      "type": "CALL",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          0,
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/delayP",
      },
      "type": "CALL",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          "https://example.com",
          Object {
            "payload": Object {
              "key": "value",
            },
          },
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/mockConstructor",
      },
      "type": "CALL",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          0,
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/delayP",
      },
      "type": "CALL",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          "https://example.com",
          Object {
            "payload": Object {
              "key": "value",
            },
          },
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/mockConstructor",
      },
      "type": "CALL",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          0,
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/delayP",
      },
      "type": "CALL",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          "https://example.com",
          Object {
            "payload": Object {
              "key": "value",
            },
          },
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/mockConstructor",
      },
      "type": "CALL",
    },
  ],
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "action": Object {
          "payload": Object {
            "message": undefined,
            "status": undefined,
          },
          "type": "DUMMY_FAILURE",
        },
        "channel": undefined,
      },
      "type": "PUT",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "action": Object {
          "payload": Object {
            "message": undefined,
            "status": undefined,
            "type": "error",
          },
          "type": "DUMMY_SHOW_ALERT",
        },
        "channel": undefined,
      },
      "type": "PUT",
    },
  ],
}
`;

exports[`retry on error cases should work with custom shouldAborts 1`] = `
Object {
  "call": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          "https://example.com",
          Object {
            "payload": Object {
              "key": "value",
            },
          },
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/mockConstructor",
      },
      "type": "CALL",
    },
  ],
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "action": Object {
          "payload": Object {
            "message": "Rejected",
            "status": 401,
          },
          "type": "DUMMY_FAILURE",
        },
        "channel": undefined,
      },
      "type": "PUT",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "action": Object {
          "payload": Object {
            "message": "Rejected",
            "status": 401,
            "type": "error",
          },
          "type": "DUMMY_SHOW_ALERT",
        },
        "channel": undefined,
      },
      "type": "PUT",
    },
  ],
}
`;

exports[`retry on success cases should execute once and dispatch the success action 1`] = `
Object {
  "call": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          "https://example.com",
          Object {
            "payload": Object {
              "key": "value",
            },
          },
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/mockConstructor",
      },
      "type": "CALL",
    },
  ],
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "action": Object {
          "payload": Object {
            "result": "Resolved",
          },
          "type": "DUMMY_SUCCESS",
        },
        "channel": undefined,
      },
      "type": "PUT",
    },
  ],
}
`;
